### PR TITLE
Refactor all of our various cra build steps

### PR DESCRIPTION
**Pull request by riknoll**
I will soon be adding another CRA for the tutorial tool, so I figured it was about time to refactor all of this build code to make it easier to add new apps and generally unify things.

@thsparks since you did this most recently, did I miss anything here?

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -863,11 +787,6 @@ exports.watch = initWatch;
 exports.watchCli = initWatchCli;
 exports.testlanguageservice = testlanguageservice;
 exports.onlinelearning = onlinelearning;
-exports.skillmap = skillmap;
-exports.authcode = authcode;
-exports.multiplayer = multiplayer;
-exports.kiosk = kiosk;
-exports.teacherTool = teacherTool;
 exports.tt = teacherTool;
```
Comment (thsparks) [inline gulpfile.js:790]: FWIW there's also `exports.tt` here (just a shortcut to teachertool), but I do actually use that on occasion when I only want to build teacher tool. Is there a reason we're removing these?

File: gulpfile.js [hunk 2]
```diff
@@ -568,109 +576,25 @@ const testSkillmap = gulp.series(buildSkillmapTests, copySkillmapTests, runSkill
...
-    .pipe(gulp.dest(`${teacherToolOut}/js`));
-
-const copyTeacherToolHtml = () => rimraf("webapp/public/teachertool.html")
-    .then(() => gulp.src(`${teacherToolRoot}/build/index.html`)
-                    .pipe(replace(/="\/static\//g, `="/blb/teachertool/`))
-                    .pipe(concat("teachertool.html"))
-                    .pipe(gulp.dest("webapp/public")));
-
-const teacherTool = gulp.series(cleanTeacherTool, buildTeacherTool, gulp.series(copyTeacherToolCss, copyTeacherToolJs, copyTeacherToolHtml));
+const teacherTool = createWebappTasks("teachertool");
```
Comment (thsparks) [inline gulpfile.js:597]: Down below here, we have the "webapp build wrapper" section which calls into the update*Strings functions. That could maybe be made slightly simpler with a unified updateWebappStrings function so we don't have to create a new one for each app, but not a huge deal.

File: package.json [hunk 3]
```diff
@@ -34,16 +34,8 @@
...
-    "built/web/authcode/css/*.css",
-    "built/web/authcode/js/*.js",
-    "built/web/multiplayer/css/*.css",
-    "built/web/multiplayer/js/*.js",
-    "built/web/kiosk/css/*.css",
-    "built/web/kiosk/js/*.js",
-    "built/web/teachertool/css/*.css",
-    "built/web/teachertool/js/*.js",
+    "built/web/*/css/*.css",
+    "built/web/*/js/*.js",
```
Comment (thsparks) [inline package.json:38]: It'd be nice if we could update the npm-prepare.js script (gets called as a part of prepare at the bottom of this file) to also reference subwebapp.ts but I'm not sure if that's possible...

**Review by thsparks — APPROVED** [review]
Huge improvement! I think only the `webappNames` issue in `server.ts` is likely to cause actual problems. Other comments are just potential opportunities for further simplification, though they may not be worth the trouble. This is a big help either way!

**Comment by thsparks** [comment]
Maybe not very helpful at this point, but FWIW, I saved a branch which just had the bare-bones updates I needed to make a new CRA. Figured I'd mention it, in case it's a useful reference: https://github.com/microsoft/pxt/compare/master...thsparks/new_react_app#diff-d7ffbddb3a36042f0033638e9e21f556c8873232b040c3f6579488d74adec3daR6

File: gulpfile.js (hunk 1 above)
Comment (riknoll) [reply gulpfile.js:790]: they're still there! just being set inside the function defining the tasks

Comment (thsparks) [reply gulpfile.js:790]: Ahh, I see now. Makes sense!

File: cli/server.ts [hunk 4]
```diff
@@ -1184,19 +1184,11 @@ export function serveAsync(options: ServeOptions) {
...
-
-        if (pathname == "/--authcode") {
-            sendFile(path.join(publicDir, 'authcode.html'));
-            return
-        }
-
-        if (pathname == "/--multiplayer") {
-            sendFile(path.join(publicDir, 'multiplayer.html'));
-            return
+        for (const subapp of SUB_WEBAPPS) {
```
[the remark this answers is not in the document]
Comment (riknoll) [reply cli/server.ts:1187]: ah, good catch. lemme see what i can do here...

[the remark this answers is not in the document]
Comment (riknoll) [reply cli/server.ts:1182]: fixed!

File: gulpfile.js (hunk 2 above)
Comment (riknoll) [reply gulpfile.js:588]: done!

File: package.json (hunk 3 above)
Comment (riknoll) [reply package.json:39]: done! I moved this config into a json file that gets used in npm-prepare.js, in the cli, and in the gulpfile

Comment (thsparks) [reply package.json:39]: Nice 😎

**Pull request merged**